### ISSUE TITLE
A Question on RM training 

### ISSUE THREAD

**Issue by DaehanKim** [openai/following-instructions-human-feedback#2]
Dear authors,
I'm an AI engineer deeply impressed by your work. Great work and thank you for sharing this!

btw, I was confused on what you meant by "train on all K-choose-2 comparisons from each prompt as a single batch element".
For my understanding, you collected K response ranks (collected from K models separately) from a prompt and made k-choose-2 comparisons out of them. These k-choose-2 samples constitutes a batch (which results in 6(K=4) to 36(K=9) batch sizes) and you do forward pass with this batch. 
But I don't get why this requires a single forward pass for each completion(meaning a single response as I see). As I understand, number of samples consumed are the same as before and the difference is in the batching scheme. I guess I'm a bit mistaken.

**Comment by SonuDixit** [comment]
Consider case $K=4$, Data = (prompt = x, completions = $\(y_0, y_1, y_2, y_3\)$ ) . Each $y_i$ is a completion. We also have preference/winner for each pair(y_i, y_j).
$r_\theta(.)$ :  reward model to be learned.
In a usual scenario, for each pair $(y_i, y_j)$, we need to make two forward passes $r_\theta(x,y_i), r_\theta(x,y_j)$, (calculate the loss), and one backprop. For each completion $y_i, k-1$ backward pass will be done. We need a total of 2*(k-choose-2) FP and k-choose-2 BP.

I think they did the following - 
1. Format comparison data such that it can be used for loss calculation. - One possible format is a list of ordered tuples [(0,1), (0,2), (3,0), (1,2),  (3,1), (2,3) ] - Here for each tuple first element is the winner. the tuple (2,3) implies $y_2$ is preferred over $y_3$.
2. Forward pass for each pair $(x,y_i)$. Now we have $[r_0, r_1, r_2, r_3]$
3. Now using the comparison tuples (1 above), calculate loss on the rewards. And then do the Backward pass. Here we need K FP, 1 BP.
